File: Moji/Views/StreamView/Stream Layouts/StreamMetrics.m
//
//  StreamMetrics.m
//  Moji
//
//  Created by Sergey Maximenko on 8/18/15.
//  Copyright (c) 2015 Ravenpod. All rights reserved.
//

#import "StreamMetrics.h"
#import "StreamView.h"

@implementation StreamMetrics

+ (instancetype)metrics:(StreamMetricsBlock)block {
    return [[[self alloc] init] change:block];
}

- (instancetype)init {
    self = [super init];
    if (self) {
        self.reusableViews = [NSMutableSet set];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super init];
    if (self) {
        self.reusableViews = [NSMutableSet set];
    }
    return self;
}

- (instancetype)change:(StreamMetricsBlock)block {
    if (block) block(self);
    return self;
}

- (BOOL)hiddenAt:(StreamIndex *)index {
    return self.hiddenBlock ? self.hiddenBlock(index, self) : self.hidden;
}

- (CGFloat)sizeAt:(StreamIndex *)index {
    return self.sizeBlock ? self.sizeBlock(index, self) : self.size;
}

- (CGRect)insetsAt:(StreamIndex *)index {
    return self.insetsBlock ? self.insetsBlock(index, self) : self.insets;
}

- (StreamReusableView*)loadView {
    
    NSMutableSet *views = _reusableViews;
    
    if (views.count > 0) {
        StreamReusableView *view = [views anyObject];
        [views removeObject:view];
        [view prepareForReuse];
        return view;
    }
    
    UINib *nib = self.nib;
    if (!nib && self.identifier) {
        nib = [UINib nibWithNibName:self.identifier bundle:nil];
    }
    if (nib) {
        NSArray *objects = [nib instantiateWithOwner:self.nibOwner options:nil];
        for (StreamReusableView *object in objects) {
            if ([object isKindOfClass:[StreamReusableView class]]) {
                object.metrics = self;
                return object;
            }
        }
    }
    return nil;
}

- (void)select:(StreamItem *)item entry:(id)entry {
    if (self.selectionBlock && entry) {
        self.selectionBlock(item, entry);
    }
}

@end

File: 16wrap/Views/StreamView/Stream Layouts/StreamMetrics.swift
//
//  StreamMetrics.swift
//  16wrap
//
//  Created by Sergey Maximenko on 9/3/15.
//  Copyright (c) 2015 Ravenpod. All rights reserved.
//

import Foundation
import UIKit

class StreamMetrics: NSObject {
    
    class func metrics(initializer: (StreamMetrics) -> Void) -> StreamMetrics {
        let metrics = self.init()
        metrics.change(initializer)
        return metrics
    }
    
    func change(initializer: (StreamMetrics) -> Void) -> StreamMetrics {
        initializer(self)
        return self
    }
    
    @IBInspectable var identifier: String?
    
    var nib: UINib?
    
    @IBOutlet weak var nibOwner:AnyObject?
    
    @IBInspectable var hidden: Bool = false
    
    var hiddenAt: (StreamIndex, StreamMetrics) -> Bool = { index, metrics in
        return metrics.hidden;
    }
    
    @IBInspectable var size: CGFloat = 0
    
    var sizeAt: (StreamIndex, StreamMetrics) -> CGFloat = { index, metrics in
        return metrics.size;
    }
    
    @IBInspectable var insets: CGRect = CGRectZero
    
    var insetsAt: (StreamIndex, StreamMetrics) -> CGRect = { index, metrics in
        return metrics.insets;
    }
    
    var selection: ((StreamItem?, AnyObject) -> Void)?
    
    var prepareAppearing: ((StreamItem?, AnyObject?) -> Void)?
    
    var finalizeAppearing: ((StreamItem?, AnyObject?) -> Void)?
    
    var reusableViews: Set<StreamReusableView> = Set()
    
    var adjustFrame = false
    
    func loadView () -> StreamReusableView? {
        
        if let view = reusableViews.first {
            reusableViews.remove(view)
            view.prepareForReuse()
            return view;
        }
        
        var nib = self.nib
        if nib == nil {
            if let identifier = self.identifier {
                nib = UINib(nibName: identifier, bundle: nil)
            }
        }
        if let objects = nib?.instantiateWithOwner(self.nibOwner, options: nil) {
            for object in objects {
                if let reusing = object as? StreamReusableView {
                    reusing.metrics = self
                    return reusing
                }
            }
        }
        
        return nil
    }
    
    func select(item:StreamItem?, entry:AnyObject) {
        if (selection != nil) {
            selection!(item, entry);
        }
    }
    
}